pywbem_mock/_wbemconnection_mock.py

```diff
@@ -995,6 +1012,104 @@ def _display_objects(obj_type, object_repo, namespace, cmt_begin, cmt_end,
                     else:
                         _uprint(dest, obj.tomof())
 
+    def register_provider(self, namespaces, classnames, provider_type,
+                          provider):
+        """
+        Register the provider object for namespace and class. Registering a
+        provider tells the FakedWBEMConnection that the provider implementation
+        provided with this call is to be executed as the request response
+        method for the namespace and class defined in lieu of the default
+        provider.
+
+        Providers can only be registered for request response methods
+        CreateInstance and DeleteInstance and InvokeMethod
+
+        Parameters:
+
+          namespace (:term:`string or :class:`py:list` of :term:`string):
+            Namespace or namespaces for which the provider is being registered.
+            If empty or None, the classname is registered for all namespaces
+
+          classname (:term:`string or :class:`py:list` of :term:`string):
+            Classname or classnames for which the provider is being
+            registered. At least one classname is required
+
+          # TODO Requirement is that it is a class that is a subclass of
+          # DefaultInstanceProvider
+          provider (:class:):
```

Comment:
I think in order to make the provider types and which operations they can implement more clear to users, we should have a separate class for each provider type. These classes have the default implementation and only those operation methods that can be overridden, plus they identify the provider type, so that a separate provider_type parameter is not needed when registering the provider. These classes can have a provider_type attribute that is set in the base class init.

Finally, I think we should name our C-U-D instance provider not just instance provider but something that expresses this subset of instance operations.

Proposal for the class names of the default provider classes user providers would inherit from:
* `InstanceWriteProvider` - C-U-D
* `MethodProvider` - InvokeMethod (both static and non-static)

Reply:
Ahha. The part you did not catch is  that I had not put the ModifyInstance into the DefaultInstanceProvider.  That done for next update.

1. Agreed that MethodProvider is a completely separate beast.

2. RE: InstanceWriteProvider.. If that is name, should it not be DefaultInstanceWriteProvider?

**DISCUSSION**

I think that the idea of provider_type in register_provider makes sense. Otherwise we need to sort that out in some other way. , Create a separate registry for each provider type. Even then, the provider_type on register_provider makes it easier to define which provider_registry would be used.

On the dispatcher side, I want to be careful until I do the method provider this afternoon.

**TODO**

Reply:
Refactoring the method_callback_interface to a method provider is left to be done in a subsequent PR.